### ISSUE TITLE
Incorrect reporting of non-exhaustive match branches of mod operation on unsigned integer

### ISSUE THREAD

**Issue by Piturnah** [rust-lang/rust#96262]
Given the following code: [play.rust-lang.org](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6cc4b13c7cb78ccc026b68e73434b3ff)
```rust
fn match_mod(num: usize) {
    match num % 3 {
	    0 => {}
	    1 => {}
	    2 => {}
	};
}
```

The current output is:

```
error[[E0004]](https://doc.rust-lang.org/stable/error-index.html#E0004): non-exhaustive patterns: `_` not covered
 [--> src/lib.rs:2:11
](https://play.rust-lang.org/#)  |
2 |     match num % 3 {
  |           ^^^^^^^ pattern `_` not covered
  |
  = help: ensure that all possible cases are being handled, possibly by adding wildcards or more match arms
  = note: the matched value is of type `usize`
  = note: `usize` does not have a fixed maximum value, so a wildcard `_` is necessary to match exhaustively
```

Ideally, this should compile.

We are in fact providing exhaustive branches, because an unsigned integer `mod n` will always be an element of the natural numbers up to `n` (not inclusive)

Tested in stable, beta, nightly.

**Comment by asquared31415** [comment]
While this would ideally work, currently match exhaustiveness is checked at the *type* level.  The expression `num % 3` produces a `usize` and so you must exhaustively match a `usize`.  I think this is by design, but still a potential improvement could be made with some major changes to that design.

**Comment by estebank** [comment]
@asquared31415 has it exactly right, this would require us to be able to specify "this `usize` is a value between `0` and `3` exclusive", which the type system isn't powerful enough to do yet. const generics could potentially get us closer to that, but I don't think the `%` operation will be able to be expanded to leverage that due to backwards compatibility.

**Comment by scottmcm** [comment]
I think I'm actually just going to close this, as expected behaviour that's unlikely to change.  Having this would essentially amount to bringing back typestate, which would be a massive change.

Note that this is *far* from the only place where Rust uses type-level logic instead of full-details-of-where-the-value-came-from logic.  This is why `while true {}` and `loop {}` have different results in type checking, why `let x; if true { x = 4; } dbg!(x);` is a "possibly uninitialized" error, etc.  (And this is handy for debugging, since you don't change borrowck by using `if true` for a bit.)

I would love to have ranged integers in the type system -- as a separate type, for the reasons @estebank implied -- but we're probably not far enough along in `const` bounds to do that yet, so it can be thought about in IRLO/Zulip, without needing an issue open for it yet.
